extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -331,12 +335,63 @@ export class AmpAnalytics extends AMP.BaseElement {
       request = addParamsToUrl(request, this.config_['extraUrlParams']);
     }
 
-    this.config_['vars']['requestCount']++;
 
+    return this.isSampledIn_(trigger).then(result => {
+      if (!result) {
+        return Promise.resolve('');
+      }
+
+      this.config_['vars']['requestCount']++;
+      request = this.expandTemplate_(request, trigger);
+
+      // For consistency with amp-pixel we also expand any url replacements.
+      return urlReplacementsFor(this.getWin()).expand(request).then(request => {
+        this.sendRequest_(request, trigger);
+        return request;
+      });
+    });
+  }
+
+  /**
+   * @param {!JSONObject} trigger The config to use to determine sampling.
+   * @return {!Promise.<boolean>} Whether the request should be sampled in or
+   * not based on sampleSpec.
+   * @private
+   */
+  isSampledIn_(trigger) {
+    const spec = trigger['sampleSpec'];
+    const resolve = Promise.resolve(true);
+    if (!spec) {
+      return resolve;
+    }
+    const threshold = spec['threshold'];
+    if (!spec['sampleOn'] ||
+        Number.isNaN(parseFloat(threshold)) || !Number.isFinite(threshold)) {
+      console./*OK*/error(this.getName_(), 'Invalid sampling spec.');
+      return resolve;
+    }
+    const key = this.expandTemplate_(spec['sampleOn'], trigger);
```

Comment:
aside: I had called the 'sampleOn' attribute the trait in my initial design. 

GA wants to consistently sample hits based on clientIds[1]. Others might want to sample based on random numbers, page path or something else. That is the reason I added the attribute 'sampleOn'.  

If we were to use getRandomValues, we lose the ability to specify what to sample based on. You suggested that we always use the same number to sample. Any ideas on how that would work? Will the random number be a trigger level thing? or a tag level thing or something else?


[1] Sampling based on clientIds results in user sessions to be consistently in the reports or out of them.